src/EditorFeatures/Core/Implementation/Classification/SemanticClassificationUtilities.cs

```diff
@@ -149,5 +209,210 @@ internal static class SemanticClassificationUtilities
+        {
+            var workspace = document.Project.Solution.Workspace;
+            var fullyLoadedStateTask = s_workspaceToFullyLoadedStateTask.GetValue(
+                workspace, w =>
+                {
+                    var workspaceLoadedService = workspace.Services.GetRequiredService<IWorkspaceStatusService>();
+                    return workspaceLoadedService.WaitUntilFullyLoadedAsync(CancellationToken.None);
+                });
+
+            // If we're not fully loaded try to read from the cache instead so that classifications appear up to date.
+            // New code will not be semantically classified, but will eventually when the project fully loads.
+            var isFullyLoaded = fullyLoadedStateTask.IsCompleted;
+            if (await TryAddSemanticClassificationsFromCacheAsync(document, textSpan, classifiedSpans, isFullyLoaded, cancellationToken).ConfigureAwait(false))
+                return;
+
+            await classificationService.AddSemanticClassificationsAsync(
+                document, textSpan, classifiedSpans, cancellationToken).ConfigureAwait(false);
+
+            // Once fully loaded, add this document to the work queue to persist its classifications to storage in the
+            // future.  This way we can store the full accurate set of classifications for it to help speed things up in
+            // VS on next launch.
+            //
+            // Only bother to do this for real documents that are being classified, not documents for things like
+            // ephemeral preview workspaces.
+            if (isFullyLoaded && document.Project.Solution.Workspace.Kind == WorkspaceKind.Host)
+                _persistClassificationsWorkQueue.AddWork(document);
+        }
+
+        private static async Task<bool> TryAddSemanticClassificationsFromCacheAsync(
+            Document document,
+            TextSpan textSpan,
+            List<ClassifiedSpan> result,
+            bool isFullyLoaded,
+            CancellationToken cancellationToken)
+        {
+            // Don't use the cache if we're fully loaded.  We should just compute values normally.
+            if (isFullyLoaded)
+                return false;
+
+            var solution = document.Project.Solution;
+            var storageService = (IChecksummedPersistentStorageService)solution.Workspace.Services.GetRequiredService<IPersistentStorageService>();
+            using var storage = storageService.GetStorage(solution);
+            if (storage == null)
+                return false;
+
+            // Try to read the existing cached data from the persistence service.  However, only bother reading it if it
+            // was stored against the same checksum that we're now checking against.
+            var checksum = await GetChecksumAsync(document, cancellationToken).ConfigureAwait(false);
+            using var stream = await storage.ReadStreamAsync(document, s_persistenceName, checksum, cancellationToken).ConfigureAwait(false);
```

Comment:
How many times does this get read during the scenario you've been looking at? I could imagine that if you open a file early, as projects load we keep retriggering the classifier which results in a bunch of duplicate reading.